src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs

```diff
@@ -2533,11 +2544,15 @@ instance ToAbstract C.LHSCore (A.LHSCore' C.Expr) where
                   "head of copattern needs to be a field identifier, but "
                   ++ prettyShow d ++ " isn't one"
         A.LHSProj (AmbQ ds) <$> toAbstract l <*> (mergeEqualPs <$> toAbstract ps2)
-    toAbstract (C.LHSWith core wps ps) = do
-      liftA3 A.LHSWith
-        (toAbstract core)
-        (toAbstract wps)
-        (toAbstract ps)
+    toAbstract (C.LHSWith core ell wps ps) = do
+      wps  <- toAbstract wps
+      ps   <- toAbstract ps
+      -- Andreas, 2020-06-02, issue #4704
+      -- Do @core@ (potentially coming from ellipsis) last because
+      -- pattern variables in (w)ps may shadow those coming from the ellipsis.
+      -- NOT NEEDED: -- when ell $ bindVarsToBind
+      core <- locallyTC eScopeCheckingEllipsis (const ell) $ toAbstract core
+      return $ A.LHSWith core wps ps
```

Comment:
I don't understand why my changes trigger `Unexpected with patterns` later in the type checker, when I have nested `with`:
```agda
f : Nat
f with zero
... | n with (suc n)
... | m = m
```
How is the abstract syntax created with my changes different from what we got before?

Reply:
I think I've found the problem. The function `numberOfWithPatterns` makes use of `foldCPattern` which uses `traverseCPatternM` in the `CPatternLike` typeclass. However, this function ignores everything that's wrapped inside an `EllipsisP`, so if there are nested `with` applications that are wrapped inside an `EllipsisP` then they are not counted.

Reply:
Wait, no nvm I see you actually updated that instance.

Reply:
To be honest I can't see anything wrong with the code, it seems the only way to find the cause is the hard way by adding debug prints along the pipeline to see where the input goes bad.